particles/Tutorial/Javascript/Demo/source/TTTGame.js

```diff
@@ -0,0 +1,131 @@
+/**
+ * @license
+ * Copyright 2019 Google LLC.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * Code distributed by Google as part of this project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+'use strict';
+
+/* global defineParticle */
+defineParticle(({SimpleParticle, html}) => {
+
+  const winningSequences = [
+    [0, 1, 2],
+    [3, 4, 5],
+    [6, 7, 8],
+    [0, 3, 6],
+    [1, 4, 7],
+    [2, 5, 8],
+    [0, 4, 8],
+    [2, 4, 6]
+  ];
+
+  const template = html`
+It is your turn <span>{{playerDetails}}</span>.
+<div slotid="gameSlot"></div>
+<div hidden="{{hideCongrats}}">Congratulations <span>{{winnerName}}</span>!</div>`;
+
+  return class extends SimpleParticle {
+
+    get template() {
+      return template;
+    }
+
+    shouldRender({gameState, playerOne, playerTwo}) {
+      return gameState && playerOne.id === 0 && playerTwo.id === 1;
+    }
+
+    update({gameState, playerOne, playerTwo, playerOneMove, playerTwoMove, events}) {
+
+      if (playerOne.id != 0) {
+        this.set('playerOne', {name: playerOne.name, avatar: playerOne.avatar, id: 0});
+      }
+
+      if (!playerTwo.id) {
+        this.set('playerTwo', {name: playerTwo.name, avatar: playerTwo.avatar, id: 1});
+      }
+
+      if (!gameState) {
+        this.newGame();
+      }
+
+      if (gameState && gameState.gameOver && events ) {
+        const reset = events.find(e => e.type === 'reset');
+        if (reset) {
+          this.newGame();
+        }
+      }
+
+      if (gameState && !gameState.gameOver) {
+        const mv = [playerOneMove.move, playerTwoMove.move][gameState.currentPlayer];
+        const avatar = [playerOne.avatar, playerTwo.avatar][gameState.currentPlayer];
+        let gs = {...gameState};
+        if (this.isMoveValid(gs, mv)) {
+          gs = this.applyMove(gs, mv, avatar);
+          gs = this.applyGameStatus(gs, playerOne, playerTwo);
+          this.set('gameState', gs);
+        }
+      }
+    }
+
+    render({gameState, playerOne, playerTwo}) {
+      return {
+        hideCongrats: !gameState.gameOver,
+        playerDetails: this.getPlayerDetails(gameState.currentPlayer, playerOne, playerTwo),
+        winnerName: gameState.winnerId !== null ? this.getPlayerDetails(gameState.winnerId, playerOne, playerTwo) : `it's a tie`
+      };
+    }
+
+    newGame() {
+      this.set('playerOneMove', {});
+      this.set('playerTwoMove', {});
+      this.clear('events');
+      this.set('gameState', {
+        board: JSON.stringify(['', '', '', '', '', '', '', '', '']),
```

Comment:
Using `JSON.stringify/parse` on demand is less robust then translating the value only when it's read/written from/to the handle.